helpers/kernel_symbols.go

```diff
@@ -48,41 +44,79 @@ type addrAndOwner struct {
 }
 
 type KernelSymbolTable struct {
-	symbols      map[name][]*KernelSymbol
-	addrs        map[addr][]*KernelSymbol
-	symByName    map[nameAndOwner][]*KernelSymbol
-	symByAddr    map[addrAndOwner][]*KernelSymbol
-	textSegStart uint64
-	textSegEnd   uint64
-	updateLock   *sync.RWMutex
-	updateWg     *sync.WaitGroup
+	symbols       map[name][]*KernelSymbol
+	addrs         map[addr][]*KernelSymbol
+	requiredSyms  map[string]struct{}
+	requiredAddrs map[uint64]struct{}
+	onlyRequired  bool
+	symByName     map[nameAndOwner][]*KernelSymbol
+	symByAddr     map[addrAndOwner][]*KernelSymbol
+	updateLock    *sync.Mutex
+	updateWg      *sync.WaitGroup
 }
 
-func NewKernelSymbolTable() (*KernelSymbolTable, error) {
+func NewKernelSymbolTable(opts ...KSymbTableOption) (*KernelSymbolTable, error) {
 	k := KernelSymbolTable{
-		updateLock: &sync.RWMutex{},
+		updateLock: &sync.Mutex{},
+	}
+	for _, opt := range opts {
+		err := opt(&k)
+		if err != nil {
+			return nil, err
+		}
 	}
+
+	k.onlyRequired = k.requiredAddrs != nil || k.requiredSyms != nil
+
 	return &k, k.Refresh()
 }
 
+type KSymbTableOption func(k *KernelSymbolTable) error
+
+func WithRequiredSymbols(reqSyms []string) KSymbTableOption {
+	return func(k *KernelSymbolTable) error {
+		k.requiredSyms = sliceToValidationMap(reqSyms)
+		return nil
+	}
+}
+
+func WithRequiredAddresses(reqAddrs []uint64) KSymbTableOption {
+	return func(k *KernelSymbolTable) error {
+		k.requiredAddrs = sliceToValidationMap(reqAddrs)
+		return nil
+	}
+}
+
 //
 // Getters (return a copy of the symbol for thread safety).
 //
 
 // TextSegmentContains returns true if the given address is in the kernel text segment.
 func (k *KernelSymbolTable) TextSegmentContains(addr uint64) (bool, error) {
-	k.updateLock.RLock()
-	defer k.updateLock.RUnlock()
+	k.updateLock.Lock()
+	defer k.updateLock.Unlock()
```

Comment:
As the goal is to squeeze as much as possible (optimise wise), what do you think unlocking right after the getTextSegmentAddresses() call?